### DISCUSSION TITLE
Exclude not required files

### DISCUSSION THREAD

**Discussion by sorenwacker** [pyinstaller/pyinstaller#6240] [Help]
Hi, I opened this question on [stackoverflow](https://stackoverflow.com/questions/69305901/how-to-prevent-pyinstaller-from-adding-unneccessary-and-unrelated-files). I noticed that some files ended up in the pyinstaller generated directory. For example, my `.git` folder and some executable files generated with InnoSetup. I don't understand why this directories and files are ending up in the generated folder as they are irrelevant for the script. I tired to exclude these files with the `exclude` arguement, but that had no effect. 

The way I am using pyinstaller is from the root directory of my package.

```
|-package/__init__.py
|   |...
|-specfiles/pyinstaller_specfile.spec
|-scripts/mainscript.py
|-hooks/...
|-setup.py
|...

```
And I call 

    pyinstaller --onedir --noconfirm specfiles\pyinstaller_specfile.spec --additional-hooks-dir=hooks

**Comment by bwoodsend** [comment]
Those files won't be collected unless you've added them as data files. Try adding a `import pprint; pprint.pprint(a.datas)` to your spec file then see if you can track down the offending entry. In particular, look for empty strings which would imply *include everything in the current working directory*.

**Comment by sorenwacker** [comment]
My datas looks like this:

```
            datas=[
                    ('C:\\Users\\soere\\anaconda3\\envs\\ms-mint\\lib\\site-packages\\dash\\favicon.ico', '.\\dash'),
                    ('C:\\Users\\soere\\anaconda3\\envs\\ms-mint\\lib\\site-packages\\dash_html_components', '.\\dash_html_components'),
                    ('C:\\Users\\soere\\anaconda3\\envs\\ms-mint\\lib\\site-packages\\dash_table', '.\\dash_table'),
                    ('C:\\Users\\soere\\anaconda3\\envs\\ms-mint\\lib\\site-packages\\dash_tabulator', '.\\dash_tabulator'),
                    ('C:\\Users\\soere\\anaconda3\\envs\\ms-mint\\lib\\site-packages\\dash_core_components', '.\\dash_core_components'),
                    ('C:\\Users\\soere\\anaconda3\\envs\\ms-mint\\lib\\site-packages\\dash_renderer', '.\\dash_renderer'),
                    ('C:\\Users\\soere\\anaconda3\\envs\\ms-mint\\lib\\site-packages\\dash_bootstrap_components', '.\\dash_bootstrap_components'),
                    ('C:\\Users\\soere\\anaconda3\\envs\\ms-mint\\lib\\site-packages\\dash_html_components', '.\\dash_html_components'),
                    ('C:\\Users\\soere\\anaconda3\\envs\\ms-mint\\lib\\site-packages\\dash_extensions', '.\\dash_extensions'),
                    ('C:\\Users\\soere\\anaconda3\\envs\\ms-mint\\lib\\site-packages\\dash_uploader', '.\\dash_uploader'),
                    ('C:\\Users\\soere\\anaconda3\\envs\\ms-mint\\lib\\site-packages\\plotly', '.\\plotly'),
                    ('C:\\Users\\soere\\anaconda3\\envs\\ms-mint\\lib\\site-packages\\pymzml', '.\\pymzml'),
                    ('C:\\Users\\soere\\anaconda3\\envs\\ms-mint\\lib\\site-packages\\bs4', '.\\bs4'),
                    ('C:\\Users\\soere\\workspace\\ms-mint\\ms_mint\\static\\ChEBI-Chem.parquet', '.\\ms_mint\\static'),                    
                    ('C:\\Users\\soere\\workspace\\ms-mint\\ms_mint\\static\\ChEBI-Groups.parquet', '.\\ms_mint\\static'),               
                    ],

```

I don't see any empty strings. A attache the output and the specfile.
 
[specfile_and_output.zip](https://github.com/pyinstaller/pyinstaller/files/7221359/specfile_and_output.zip)

Changing the directory solved it partly. Apparently, everything that is in the same directory as where the command is executed is integrated. When I do that once, a `build` and a `dist` folder are generated. When I run it a second time, the dist folder will contain a copy of the `dist` and `build` folders. 

Even when I clean the current working directory the dist folder will include a copy of the build folder. 

Thank you for your time.

**Comment by sorenwacker** [comment]
Looks like this is due to an outdated version of pyinstaller.

**Comment by sorenwacker** [comment]
But now I am getting this error...
```

    from . import add_metab
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "PyInstaller\loader\pyimod03_importers.py", line 546, in exec_module
  File "ms_mint\app\add_metab.py", line 16, in <module>
    CHEBI_CHEM_FN = os.path.abspath(resource_filename('ms_mint.static', 'ChEBI-Chem.parquet'))
  File "pkg_resources\__init__.py", line 1130, in resource_filename
  File "pkg_resources\__init__.py", line 349, in get_provider
  File "pkg_resources\__init__.py", line 1379, in __init__
  File "ntpath.py", line 223, in dirname
  File "ntpath.py", line 185, in split
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

**Reply by sorenwacker** [reply]
Using setuptools-58.1.0

**Reply by rokm** [reply]
I think this is caused by `ms_mint.static` module not being collected (as it is not directly imported anywhere in the code; only its resources are queried). In my test environment, I'm getting a module-not found error, though.

Anyway, I've cleaned up your spec file and refactored it into hooks (so we can upstream them to the hooks-contrib repository) in sorenwacker/ms-mint#38. I've tested the build using a clean virtual environment with: python 3.8.7 from python.org, PyInstaller 4.5.1, and dependencies from `requirements.txt` at versions that are available today (as most of them are unpinned). The frozen program seems to work now.

**Comment by sorenwacker** [comment]
Wow, that looks a lot cleaner. I am going to acknowledge for that in the repository. If that is OK.
